### PR TITLE
Bugfix/cam plotting memory issue

### PR DESCRIPTION
**Pull request by MarcelCaron-NOAA**
## Pull Request Testing ##

- [x] Describe testing already performed for this Pull Request:</br>
The cam/grid2obs and /precip prep, stats, and plots steps were run successfully with the bugfix. 
- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
I recommend running the jobs from the cron like this (replacing {USER} with your username and {HOMEevs} with the path to the EVS directory that is being tested):
```
30 0,6,12,18 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswarw_precip_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_hireswarw_precip_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarw_precip_prep.out 2>&1
30 0,6,12,18 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswarwmem2_precip_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_hireswarwmem2_precip_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarwmem2_precip_prep.out 2>&1
30 0,6,12,18 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswfv3_precip_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_hireswfv3_precip_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswfv3_precip_prep.out 2>&1
30 0,6,12,18 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hrrr_precip_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_hrrr_precip_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hrrr_precip_prep.out 2>&1
30 0,6,12,18 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_namnest_precip_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_namnest_precip_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_namnest_precip_prep.out 2>&1
0 19-22 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswarw_precip_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_hireswarw_precip_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarw_precip_stats.out 2>&1
0 19-22 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswarwmem2_precip_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_hireswarwmem2_precip_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarwmem2_precip_stats.out 2>&1
0 19-22 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswfv3_precip_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_hireswfv3_precip_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswfv3_precip_stats.out 2>&1
0 19-22 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hrrr_precip_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_hrrr_precip_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hrrr_precip_stats.out 2>&1
0 19-22 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_namnest_precip_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_namnest_precip_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_namnest_precip_stats.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswarw_grid2obs_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_hireswarw_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarw_grid2obs_prep.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswarwmem2_grid2obs_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_hireswarwmem2_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarwmem2_grid2obs_prep.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswfv3_grid2obs_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_hireswfv3_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswfv3_grid2obs_prep.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hrrr_grid2obs_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_hrrr_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hrrr_grid2obs_prep.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_namnest_grid2obs_prep.out {HOMEevs}/ecf/scripts/cam/prep/jevs_namnest_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_namnest_grid2obs_prep.out 2>&1
0 2,3,6,9,12,15,18,21 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswarw_grid2obs_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_hireswarw_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarw_grid2obs_stats.out 2>&1
0 2,3,6,9,12,15,18,21 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswarwmem2_grid2obs_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_hireswarwmem2_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarwmem2_grid2obs_stats.out 2>&
0 2,3,6,9,12,15,18,21 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswfv3_grid2obs_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_hireswfv3_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswfv3_grid2obs_stats.out 2>&1
0 2,3,6,9,12,15,18,21 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hrrr_grid2obs_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_hrrr_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hrrr_grid2obs_stats.out 2>&1
0 2,3,6,9,12,15,18,21 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_namnest_grid2obs_stats.out {HOMEevs}/ecf/scripts/cam/stats/jevs_namnest_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_namnest_grid2obs_stats.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_cam_grid2obs_plots.out {HOMEevs}/ecf/scripts/cam/plots/jevs_cam_grid2obs_plots.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_cam_grid2obs_plots.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_cam_precip_plots.out {HOMEevs}/ecf/scripts/cam/plots/jevs_cam_precip_plots.ecf >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_cam_precip_plots.out 2>&1
```

The prep jobs should take about 1-2 minutes to complete. The stats jobs shouldn't take more than 2 hours each to complete. The plots jobs take about an hour to complete.

I can check for clean runs and a successfully outputted plots tarball for each of the two verif cases.

- [x] Has the code been checked to ensure that no errors occur during the execution? 
Yes
- [x] Do these updates/additions include sufficient testing updates? 
Yes
- [x] Please complete this pull request review by 6/23/2023.</br>

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

### CODE REVIEW

**Comment by ShelleyMelchior-NOAA** [comment]
Just FYI that I have cloned your fork and checked out `bugfix/cam_plotting_memory_issue`. I will begin working on the PR soon.

**Comment by ShelleyMelchior-NOAA** [comment]
@MarcelCaron-NOAA  -- I set up the crons for grid2obs and precip for prep, stats, and plots.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0`
output logs: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out`

I ran stats and plots pointing to my local prep and stats. Let me know if I should be using the EVS para prep and stats instead.

**Comment by MarcelCaron-NOAA** [comment]
All grid2obs prep and stats output look good to me.  The precip prep were successful so far today, so I'll check the stats output at the end of the day.  The stats jobs are fine to point to the local directories.

The plots jobs should actually point to the EVS para stats directory.  Also, I forgot to note that both grid2obs and precip plots jobs may run at any time of the day even though I set them to 0Z in cron.

**Comment by ShelleyMelchior-NOAA** [comment]
Thanks, @MarcelCaron-NOAA  -- I will re-run the plots precip and grid2obs jobs pointing to the EVS parallel stats files. I will re-run interactively since the jobs can run any time. I will provide another update when complete.

File: ecf/scripts/cam/plots/jevs_cam_grid2obs_plots.ecf [hunk 1]
```diff
@@ -60,32 +59,29 @@ module load zlib/${zlib_ver}
...
-export COMINccpa=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT/$RUN
-export COMINmrms=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT/$RUN
-export COMINspcotlk=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT/$RUN
-export COMIN=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/stats/$COMPONENT/$RUN
-export COMOUT=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/$STEP/$COMPONENT/$RUN
+export COMINccpa=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT
+export COMINmrms=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT
+export COMINspcotlk=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT
+export COMIN=/lfs/h2/emc/vpppg/noscrub/${USER}/$NET/$evs_ver/stats/$COMPONENT
+export COMOUT=/lfs/h2/emc/vpppg/noscrub/${USER}/$NET/$evs_ver/$STEP/$COMPONENT
```
Comment (ShelleyMelchior-NOAA) [inline ecf/scripts/cam/plots/jevs_cam_grid2obs_plots.ecf]: Remove `$STEP/$COMPONENT` from line 84 b/c $COMOUT is used in the definition for $COMOUTplots in `JEVS_CAM_PLOTS`: 
`export COMOUTplots=$COMOUT/$STEP/$COMPONENT/$RUN.${VDATE}`

File: ecf/scripts/cam/plots/jevs_cam_precip_plots.ecf [hunk 2]
```diff
@@ -60,32 +59,29 @@ module load zlib/${zlib_ver}
...
-export COMINccpa=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT/$RUN
-export COMINmrms=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT/$RUN
-export COMINspcotlk=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT/$RUN
-export COMIN=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/stats/$COMPONENT/$RUN
-export COMOUT=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/$STEP/$COMPONENT/$RUN
+export COMINccpa=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT
+export COMINmrms=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT
+export COMINspcotlk=/lfs/h2/emc/ptmp/${USER}/EVS_out/com/$NET/$evs_ver/prep/$COMPONENT
+export COMIN=/lfs/h2/emc/vpppg/noscrub/${USER}/$NET/$evs_ver/stats/$COMPONENT
+export COMOUT=/lfs/h2/emc/vpppg/noscrub/${USER}/$NET/$evs_ver/$STEP/$COMPONENT
```
Comment (ShelleyMelchior-NOAA) [inline ecf/scripts/cam/plots/jevs_cam_precip_plots.ecf]: Remove `$STEP/$COMPONENT` from line 84 b/c $COMOUT is used in the definition for $COMOUTplots in `JEVS_CAM_PLOTS`:
`export COMOUTplots=$COMOUT/$STEP/$COMPONENT/$RUN.${VDATE}`

**Review by ShelleyMelchior-NOAA — CHANGES_REQUESTED** [review]
Re-running plots pointing to the EVS para stats files has revealed one issue in the plots ecf scripts. Please correct as indicated.

**Comment by MarcelCaron-NOAA** [comment]
Thanks, I've corrected the scripts to use COMOUTplots, and then changed the path used to define COMOUT in the ecf scripts.

**Comment by MarcelCaron-NOAA** [comment]
All of the jobs completed as expected except for the plotting jobs.  The grid2obs plotting job completed successfully, but took 5 hours to complete.  The precip plotting job failed due to a memory issue.  I am running this on Dogwood this evening, and will try again on Cactus tomorrow to rule out that this may be a temporary issue with WCOSS2.

**Comment by ShelleyMelchior-NOAA** [comment]
@MarcelCaron-NOAA , I managed to get another plots run in on dogwood last night before going to bed.
I can get grid2obs to work.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/plots/cam/atmos.20230620/` (dogwood)
I did not untar the file. If you cannot access dogwood I can cp the tar file to /dfs/read on cactus for you to retrieve.

I think the precip job ended early b/c of memory issues, still. I guess this is a dogwood fluke. I will try to run again on cactus once it is available later today.

**Comment by ShelleyMelchior-NOAA** [comment]
@MarcelCaron-NOAA -- I re-ran cam_precip plots on cactus. The job finished early (after 1:42) in error. See the output log file ~line 59,799. 
`/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_precip_plots.o65316680`

Let me know if perhaps I don't have something set up correctly.

**Comment by MarcelCaron-NOAA** [comment]
Thanks @ShelleyMelchior-NOAA.  Not having reviewed the output I am confident that there is a memory issue, since that is what my parallel tests are experiencing.  I suspect that prior to the PR I had not been testing with emc.vpppg as I originally thought, and that archive is now quite large.  

I have an idea for reducing the memory costs, will implement and test this tonight, and will get back to you tomorrow if all goes well!

**Comment by MarcelCaron-NOAA** [comment]
@ShelleyMelchior-NOAA I've been able to fix and test the memory issue.  Sorry about the wait!  The precip plotting job should complete cleanly, and the expectation is 4 hours to complete because it is pointing to the large emc.vpppg stats archive.

I've also included the time-sensitive changes explained in Geoff's email yesterday.  I suppose it is good to test those changes, if you agree, in which case each prep, stats, and plots job listed for testing this PR may be run as originally instructed.

**Comment by MarcelCaron-NOAA** [comment]
*Amendment to the original instructions is the location of the "ecf" scripts, which are now "sh" scripts located in $HOMEevs/dev/drivers/scripts/$COMPONENT/$STEP

**Comment by ShelleyMelchior-NOAA** [comment]
Awesome, @MarcelCaron-NOAA ! I will pull in the changes to my clone of your fork and re-test everything.

**Comment by ShelleyMelchior-NOAA** [comment]
All prep processing is complete.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/cam`
log files: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_*_[grid2obs][precip]_prep.o*`

**Comment by MarcelCaron-NOAA** [comment]
The prep jobs look good.  All of them completed successfully except for the jevs_hireswarwmem2_precip_prep.sh job, but I think this is due to the HOMEevs test directory name:
/lfs/h2/emc/vpppg/noscrub/shelley.melchior/_**ghithubwkspc**_/forks/MC/EVS

should be:
/lfs/h2/emc/vpppg/noscrub/shelley.melchior/**_githubwkspc_**/forks/MC/EVS

This issue shouldn't prevent the stats jobs from running successfully.

**Comment by ShelleyMelchior-NOAA** [comment]
Thanks for finding my typo!

**Comment by ShelleyMelchior-NOAA** [comment]
@MarcelCaron-NOAA  -- I was having trouble today getting my cron entries to work today for your PR testing, so I ran things manually. But I rushed things and made mistakes, like you saw. Anyway, the cron entries appear to be running now. I will let this run over the weekend. Feel free to check COMOUT and log files.
COMOUT= `/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/*/cam/`
logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out`

**Comment by ShelleyMelchior-NOAA** [comment]
It seem the cron entries still aren't working correctly. hmmm

**Comment by MarcelCaron-NOAA** [comment]
@ShelleyMelchior-NOAA Thanks for managing the PR this weekend!  It seems the cron jobs have been working lately as I can see the output files of recent runs in your output directory.  The prep jobs are already confirmed, and I can see that the stats jobs have also all completed successfully.  

The plots jobs failed at the very end because the "COMOUTplots" definition was removed from the J-job in a recent merge to develop.  I opted to remove references to COMOUTplots and revert back to using COMOUT as the final output location for plots tarballs.  The updates were pushed to this branch.  I'll check the output logs of the next iteration(s).  Since this was a minor issue occurring after the primary memory issue was successfully mitigated, I fully expect them to finish successfully!

**Comment by ShelleyMelchior-NOAA** [comment]
Thanks Marcel! I pulled your changes into my clone of your `bugfix/cam_plotting_memory_issue`. I will let the crons continue to run and we will see what the next plots cycle looks like after 0305Z when the jobs kick off.

**Comment by ShelleyMelchior-NOAA** [comment]
W/ @MarcelCaron-NOAA 's suggestion, I have submitted the plots jobs interactively.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/plots/cam`
logfiles:
`/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_grid2obs_plots.o66968984`
`/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_precip_plots.o66968993`

**Comment by MarcelCaron-NOAA** [comment]
Hmm, both interactive jobs failed just after writing job cards.  

It looks like as the later job cards were written, the first 250 job cards were subsequently scrubbed, so that by the time the first group of job cards was submitted, the paths to those jobs cards no longer existed.  Both interactive jobs also took unusually long to run (~6 hours just to finish writing the job cards ...), leading me to believe this is a WCOSS2 issue.  

I'll run test jobs myself overnight to confirm.

**Comment by MarcelCaron-NOAA** [comment]
Precip plotting jobs have been failing because they were looking for the wrong variable name in netCDF files that are used to plot spatial maps.   I changed the variable name and will test again on Cactus tonight.

**Comment by ShelleyMelchior-NOAA** [comment]
I just re-submitted the plots processing (~10:40 EDT) now that the emc.vpppg para stats data on dogwood have caught up after the machine switch.
logfiles:
`/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_grid2obs_plots.o80932416`
`/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_precip_plots.o80931979`

**Comment by ShelleyMelchior-NOAA** [comment]
Plots processing completed for precip. @MarcelCaron-NOAA confirmed in separate chat that precip looks good.
COMOUT= `/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/plots/cam`
log file: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_precip_plots.o80931979`

**Comment by ShelleyMelchior-NOAA** [comment]
Plots processing completed for grid2obs. @MarcelCaron-NOAA , please rview.
COMOUT= `/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/plots/cam`
log file: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_grid2obs_plots.o80932416`

**Comment by MarcelCaron-NOAA** [comment]
The grid2obs plotting job completed cleanly and the output tarball is as expected in size and contents.  Same for the precip plotting job.  I'm fine with merging and closing this PR!

This branch is also now up to date with develop and should be able to be merged.  Let me know if there are any further issues

**Review by ShelleyMelchior-NOAA — APPROVED** [review]
Code changes are solid and as expected. Approved.

**Pull request merged**